Mathlib/Analysis/MeanInequalities.lean

```diff
@@ -253,6 +267,58 @@ end Real
 
 end GeomMeanLEArithMean
 
+section HarmMeanLEGeomMean
+
+/-! ### HM-GM inequality -/
+
+namespace Real
+
+/-- **HM-GM inequality**: The harmonic mean is less than or equal to the geometric mean, weighted
+version for real-valued nonnegative functions. -/
+theorem harm_mean_le_geom_mean_weighted (w z : ι → ℝ) (hs : s.Nonempty) (hw : ∀ i ∈ s, 0 < w i)
+    (hw' : ∑ i in s, w i = 1) (hz : ∀ i ∈ s, 0 < z i) :
+    1/(∑ i in s, w i / z i) ≤ ∏ i in s, z i ^ w i  := by
+    have : ∏ i in s, (1 / z) i ^ w i ≤ ∑ i in s, w i * (1 / z) i :=
+      geom_mean_le_arith_mean_weighted s w (1/z) (fun i hi ↦ le_of_lt (hw i hi)) hw'
+                                                 (fun i hi ↦ one_div_nonneg.2 (le_of_lt (hz i hi)))
+    have p_pos : 0 < ∏ i in s, (z i)⁻¹ ^ w i :=
+      prod_pos fun i hi => rpow_pos_of_pos (inv_pos.2 (hz i hi)) _
+    have s_pos : 0 < ∑ i in s, w i * (z i)⁻¹ :=
+      sum_pos (fun i hi => Real.mul_pos (hw i hi) (inv_pos.2 (hz i hi))) hs
+    norm_num at this
+    rw [← inv_le_inv s_pos p_pos, inv_eq_one_div] at this
+    apply le_trans this
+    have p_pos₂ : 0 < (∏ i in s, (z i) ^ w i)⁻¹ :=
+      inv_pos.2 (prod_pos fun i hi => rpow_pos_of_pos ((hz i hi)) _ )
+    rw [← inv_inv (∏ i in s, z i ^ w i), inv_le_inv p_pos p_pos₂, ← Finset.prod_inv_distrib]
+    gcongr
+    · refine fun i hi ↦ inv_nonneg.mpr (Real.rpow_nonneg (le_of_lt (hz i hi)) _)
+    · rw [Real.inv_rpow]
+      apply (fun i hi ↦ le_of_lt (hz i hi)); assumption
+
+
+/-- **HM-GM inequality**: The **harmonic mean is less than or equal to the geometric mean. --/
+theorem harm_mean_le_geom_mean {ι : Type*} (s : Finset ι) (hs : Finset.Nonempty s) (w : ι → ℝ)
+    (z : ι → ℝ) (hw : ∀ i ∈ s, 0 < w i) (hw' : 0 < ∑ i in s, w i) (hz : ∀ i ∈ s, 0 < z i) :
+    (∑ i in s, w i)/(∑ i in s, w i / z i) ≤ (∏ i in s, z i ^ w i) ^ (∑ i in s, w i)⁻¹ := by
+  have := harm_mean_le_geom_mean_weighted s (fun i => (w i) / ∑ i in s, w i) z hs ?_ ?_ hz
+  simp at this
```

Comment:
Since you are in the middle of the proof, you should squeeze the simp to improve the robustness, by doing `simp? at this` and using the suggested output.

Also, before this line, you have three goals, so you should have started right away with a focusing dot (not relevant if you have used `suffices`).